### PR TITLE
Reject asset upload with invalid `id`

### LINKED ISSUE #209: API should reject asset upload when `id` contain invalid characters
**Describe the bug**

For example, uploading a dataset YAML with `id: codenet_langclass` will cause issues when the PVC is created for that dataset later on, since Kubernetes resource IDs cannot contain underscore (`_`) characters (among others).

DataShim response when running the PVC mounting pipeline for a dataset with `id: codenet_langclass`:
```
HTTP response body: {
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {},
  "status": "Failure",
  "message": "Dataset.com.ie.ibm.hpsys \"codenet_langclass\" is invalid: metadata.name: Invalid value: \"codenet_langclass\": a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')",
  "reason": "Invalid",
  "details": {
    "name": "codenet_langclass",
    "group": "com.ie.ibm.hpsys",
    "kind": "Dataset",
    "causes": [
      {
        "reason": "FieldValueInvalid",
        "message": "Invalid value: \"codenet_langclass\": a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')",
        "field": "metadata.name"
      }
    ]
  },
  "code": 422
}
```

**Additional information:**

The actual `id` of the dataset asset in MLX was `codenet-langclass`, with correctly replaced underscore `_` character. However, the component generating the DataShim metadata used the original `id` field from the YAML file, with the `id` containing the underscore character.

FYI, here is a quick test of the `sanitize_k8s_name` function:

```Python
python -c "
from kfp_tekton.compiler._k8s_helper import sanitize_k8s_name; 
print(sanitize_k8s_name('under_score-dash'));
"
```
```
under-score-dash
```

**Proposed fix:**

Any `_upload_...` method in `api/server/swagger_server/controllers_impl/XXX_service_controller_impl.py` which takes an `id` (or similar field like `model_identifier`) from the uploaded YAML file to generate the asset `id` ...

api/server/swagger_server/controllers_impl/
- `dataset_service_controller_impl.py`, line 372: `def _upload_dataset_yaml(...)`
- `model_service_controller_impl.py`, line 381: `def _upload_model_yaml(...)`
- `notebook_service_controller_impl.py`, line 392: `def _upload_notebook_yaml(...)` (still being worked on #212)

... should call a new `validate_id()` method similar to the `swagger_server.controllers_impl.validate_parameters` function in `api/server/swagger_server/controllers_impl/__init__.py`:

```Python
def validate_id(id: str) -> (str, int):

    if id != sanitize_k8s_name(id)
        return f"Identifiers must contain lower case alphanumeric characters or '-' only.", 422

    return None, 200
```

### PR DESCRIPTION
**Pull request by JAulet**
**Excerpt from issue #209:**

The actual `id` of the dataset asset in MLX was `codenet-langclass`. Notice the dash `-` which correctly replaced the underscore `_` character. However, the component which generated the DataShim metadata used the value of the `id` field from the YAML file, with the `id` containing the underscore character.

**Proposed fix:**

Any `_upload_XXX` method in `api/server/swagger_server/controllers_impl/XXX_service_controller_impl.py` which takes an `id` (or similar field like `model_identifier`) from the uploaded YAML file to generate the asset `id` should call a new `validate_id()` method similar to the `swagger_server.controllers_impl.validate_parameters` function in `api/server/swagger_server/controllers_impl/__init__.py`:

`api/server/swagger_server/controllers_impl/`:
- [x] `__init__.py`, add `validate_id(id: str)` method
- [x] `dataset_service_controller_impl.py`, line 372: `def _upload_dataset_yaml(...)` ... use `validate_id(id: str)`
- [x] `model_service_controller_impl.py`, line 381: `def _upload_model_yaml(...)`... use `validate_id(id: str)`
- [x] `notebook_service_controller_impl.py`, line 392: `def _upload_notebook_yaml(...)` ... use `validate_id(id: str)`

Resolves #209

### CODE REVIEW

File: api/server/swagger_server/controllers_impl/notebook_service_controller_impl.py [hunk 1]
```diff
@@ -396,6 +396,11 @@ def _upload_notebook_yaml(yaml_file_content: AnyStr, name=None, access_token=Non
     requirements = yaml_dict["implementation"]["github"].get("requirements")
     filter_categories = yaml_dict.get("filter_categories") or dict()
 
+    errors, status = validate_id(notebook_id)
```
Comment (ckadner) [inline api/server/swagger_server/controllers_impl/notebook_service_controller_impl.py:399]: let's move this check above line 391 where we generate the `notebook_id` and validate the `yaml_dict.get("id")`

File: api/server/swagger_server/controllers_impl/model_service_controller_impl.py [hunk 2]
```diff
@@ -397,6 +397,11 @@ def _upload_model_yaml(yaml_file_content: AnyStr, name=None, existing_id=None):
     if type(api_model.servable_tested_platforms) == str:
         api_model.servable_tested_platforms = api_model.servable_tested_platforms.replace(", ", ",").split(",")
 
+    errors, status = validate_id(api_model.id)
```
Comment (ckadner) [inline api/server/swagger_server/controllers_impl/model_service_controller_impl.py:400]: let's move this check to line 375 after we load the yaml and before we create the `ApiModel` object and validate the `model_def.get("model_identifier")`

File: api/server/swagger_server/controllers_impl/dataset_service_controller_impl.py [hunk 3]
```diff
@@ -371,6 +371,11 @@ def _upload_dataset_yaml(yaml_file_content: AnyStr, name=None, existing_id=None)
     # if yaml_dict.get("id") != dataset_id:
     #     raise ValueError(f"Dataset.id contains non k8s character: {yaml_dict.get('id')}")
 
+    errors, status = validate_id(dataset_id)
```
Comment (ckadner) [inline api/server/swagger_server/controllers_impl/dataset_service_controller_impl.py:374]: let's move this check above line 366after we load the YAML and before we generate the `dataset_id` and validate the `yaml_dict.get("id")`

**Review by ckadner — CHANGES_REQUESTED** [review]
Thanks @JAulet -- the original problem was not that the `id` we generate was invalid, but the `id` field inside the YAML could still be invalid. And since that YAML itself was used by another component to get the `id` the DataShim logic failed. So this PR should add some code to verify the `id`s in the YAML adhere to the k8s identifier format.

**Comment by JAulet** [comment]
@ckadner Please Review.

**Comment by ckadner** [comment]
> @ckadner Please Review.

@JAulet -- did you try this out by uploading a model, dataset, notebook YAML with invalid ID values?

**Comment by JAulet** [comment]
> > @ckadner Please Review.
> 
> @JAulet -- did you try this out by uploading a model, dataset, and notebook YAML respectively with invalid ID values (and with valid ones)?

Yes, I did. I got the correct output for each.

**Review by ckadner — APPROVED** [review]
/lgtm

Thanks @JAulet

**Pull request merged**